File: tests/python/test_0001_two_sum.py
"""
Tests for Problem 0001: Two Sum

These tests validate the solution against known inputs and edge cases.
The primary validation is a successful LeetCode submission; these tests
provide a local feedback loop and catch regressions.
"""

import sys
import os

# Add the solution directory to the path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../problems/0001-two-sum/python'))

from solution import Solution


class TestTwoSum:
    def setup_method(self):
        self.solution = Solution()

    # --- Basic examples from LeetCode ---

    def test_example_1(self):
        """nums = [2,7,11,15], target = 9 -> [0,1]"""
        result = self.solution.twoSum([2, 7, 11, 15], 9)
        assert sorted(result) == [0, 1]

    def test_example_2(self):
        """nums = [3,2,4], target = 6 -> [1,2]"""
        result = self.solution.twoSum([3, 2, 4], 6)
        assert sorted(result) == [1, 2]

    def test_example_3(self):
        """nums = [3,3], target = 6 -> [0,1]"""
        result = self.solution.twoSum([3, 3], 6)
        assert sorted(result) == [0, 1]

    # --- Edge cases ---

    def test_negative_numbers(self):
        """Target achieved with negative numbers."""
        result = self.solution.twoSum([-1, -2, -3, -4, -5], -8)
        assert sorted(result) == [2, 4]

    def test_mixed_positive_negative(self):
        """One positive and one negative number sum to target."""
        result = self.solution.twoSum([-3, 4, 3, 90], 0)
        assert sorted(result) == [0, 2]

    def test_zero_target(self):
        """Target is zero with complementary elements."""
        result = self.solution.twoSum([0, 4, 3, 0], 0)
        assert sorted(result) == [0, 3]

    def test_large_numbers(self):
        """Numbers near integer boundaries."""
        result = self.solution.twoSum([2**30, 2**30], 2**31)
        assert sorted(result) == [0, 1]

    def test_two_elements(self):
        """Minimum valid input size."""
        result = self.solution.twoSum([1, 2], 3)
        assert sorted(result) == [0, 1]

File: tests/c/test_0001_two_sum.c
/*
 * Tests for Problem 0001: Two Sum (C)
 *
 * Compile and run:
 *   gcc -Wall -Wextra -o test_0001 test_0001_two_sum.c && ./test_0001
 *
 * This file includes the solution directly to avoid complex build setups.
 * The primary validation is a successful LeetCode submission.
 */

#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

/* Include the solution */
#include "../../problems/0001-two-sum/c/solution.c"

static int tests_passed = 0;
static int tests_failed = 0;

static void assert_two_sum(int *nums, int nums_size, int target,
                           int expected_a, int expected_b, const char *name) {
    int return_size;
    int *result = twoSum(nums, nums_size, target, &return_size);

    if (result == NULL) {
        printf("  FAIL: %s (returned NULL)\n", name);
        tests_failed++;
        return;
    }

    /* Sort the result pair for comparison */
    int ra = result[0] < result[1] ? result[0] : result[1];
    int rb = result[0] < result[1] ? result[1] : result[0];
    int ea = expected_a < expected_b ? expected_a : expected_b;
    int eb = expected_a < expected_b ? expected_b : expected_a;

    if (ra == ea && rb == eb) {
        printf("  PASS: %s\n", name);
        tests_passed++;
    } else {
        printf("  FAIL: %s (expected [%d,%d], got [%d,%d])\n",
               name, ea, eb, ra, rb);
        tests_failed++;
    }

    free(result);
}

int main(void) {
    printf("Testing Problem 0001: Two Sum (C)\n\n");

    /* Example 1 */
    {
        int nums[] = {2, 7, 11, 15};
        assert_two_sum(nums, 4, 9, 0, 1, "example 1");
    }

    /* Example 2 */
    {
        int nums[] = {3, 2, 4};
        assert_two_sum(nums, 3, 6, 1, 2, "example 2");
    }

    /* Duplicate values */
    {
        int nums[] = {3, 3};
        assert_two_sum(nums, 2, 6, 0, 1, "duplicate values");
    }

    /* Negative numbers */
    {
        int nums[] = {-1, -2, -3, -4, -5};
        assert_two_sum(nums, 5, -8, 2, 4, "negative numbers");
    }

    /* Minimum input */
    {
        int nums[] = {1, 2};
        assert_two_sum(nums, 2, 3, 0, 1, "minimum input");
    }

    printf("\nResults: %d passed, %d failed\n", tests_passed, tests_failed);
    return tests_failed > 0 ? 1 : 0;
}
